shop.cpp

```diff
@@ -3,24 +3,57 @@
 
 Shop::Shop(int mMaxChairs)
     :mMaxChairs(mMaxChairs)
-    ,mTheBarber(this)
-    ,mBarberSemaphore(0)
-    ,mCustomersSemaphore(0)
 {
+    mBarberSemaphore = new Semaphore(1);
+    mCustomersSemaphore = new Semaphore(0);
+    mTheBarber = new Barber(this, mBarberSemaphore, mCustomersSemaphore);
+}
+
+Shop::~Shop()
+{
+    if (mTheBarber != nullptr)
```

Comment:
You don't need to check pointers against null when calling [delete](http://en.cppreference.com/w/cpp/memory/new/operator_delete).
> In all cases, if ptr is a null pointer, the standard library deallocation functions do nothing

Reply:
:ok: Done